[WebAudio][GStreamer] Use single non-interleaved appsrc instead of multi appsrcs with audiointerleave

Bastian-Krause opened the pull request:
#### 63076adef16d7f28845cbedde25529bbec9b5917
<pre>
[WebAudio][GStreamer] Use single non-interleaved appsrc instead of multi appsrcs with audiointerleave
<a href="https://bugs.webkit.org/show_bug.cgi?id=247736">https://bugs.webkit.org/show_bug.cgi?id=247736</a>

Reviewed by Philippe Normand.

GStreamer 1.16 brought support for planar (non-interleaved) raw audio.
Let&apos;s use that to simplify the &quot;1 appsrc per channel + audiointerleave&quot;
approach to a single appsrc with non-interleaved layout.

We do not need to keep the appsrcs and mapped buffers per channel around
which lets us drop a bunch of code.

This approach cuts the overhead the GStreamer pipeline causes on the
WPEWebProcess in half on an NXP i.MX 6 DualLite.

* Source/WebCore/platform/audio/gstreamer/WebKitWebAudioSourceGStreamer.cpp:
(getGStreamerAudioCaps): Return caps with actual number of channels and
non-interleaved layout.
(webKitWebAudioSrcConstructed): Drop audiointerleave element, create
single appsrc.
(webKitWebAudioSrcAllocateBuffers): Drop loop over multiple sources, no
need to keep mapped buffers per channel around.
(webKitWebAudioSrcRenderAndPushFrames): Drop loop over multiple sources.
(webKitWebAudioSrcRenderIteration): Operate on single buffer.
(webKitWebAudioSrcChangeState): Drop check for audiointerleave element,
adjust buffer size calculation.
(getGStreamerMonoAudioCaps): Deleted. Replaced by getGStreamerAudioCaps.

Canonical link: <a href="https://commits.webkit.org/256648@main">https://commits.webkit.org/256648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a777a97ed33d503e41714bd51b730a3604d4bbde

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96415 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5669 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29482 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105958 "Built successfully") | [💥 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166307 "An unexpected error occured. Recent messages:Pull request contains relevant changes; Failed to print configuration") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100398 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5843 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34416 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88794 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102684 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102086 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4353 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83017 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31332 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86195 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88085 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74225 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40145 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37819 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20965 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4614 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/3403 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/823 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40234 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/audio/gstreamer/WebKitWebAudioSourceGStreamer.cpp:
```diff
@@ -117,42 +115,12 @@ static void webKitWebAudioSrcGetProperty(GObject*, guint propertyId, GValue*, GP
-        "layout", G_TYPE_STRING, "interleaved", nullptr);
-}
-
-static GstAudioChannelPosition webKitWebAudioGStreamerChannelPosition(int channelIndex)
```
Why is this removed?

Bastian-Krause replied:
It's not used anymore, see commit message.

philn replied:
How are channel positions mapped to GStreamer layout then? By magic?

Bastian-Krause replied:
Ah, now I get what you mean. Let me check how this can work.

Bastian-Krause replied:
I've updated the PR to reflect the channel position mapping.

philn reviewed (commented):
The element could be rewritten as a basesrc sub-class now, maybe based on pushsrc. Could be done later though.

philn commented on Source/WebCore/platform/audio/gstreamer/WebKitWebAudioSourceGStreamer.cpp:
```diff
@@ -155,6 +145,25 @@ static GstAudioChannelPosition webKitWebAudioGStreamerChannelPosition(int channe
+static GstCaps* getGStreamerAudioCaps(float sampleRate, unsigned numberOfChannels)
+{
+    guint64 channel_mask = 0;
+    GstAudioChannelPosition* positions = g_new(GstAudioChannelPosition, numberOfChannels);
```
`fastMalloc` is preferred in WebCore, but here maybe you could use a `Vector<GstAudioChannelPosition>` ?

Bastian-Krause replied:
Sure, I've pushed a variant with `Vector<GstAudioChannelPosition>` now.

philn commented on Source/WebCore/platform/audio/gstreamer/WebKitWebAudioSourceGStreamer.cpp:
```diff
@@ -155,6 +145,25 @@ static GstAudioChannelPosition webKitWebAudioGStreamerChannelPosition(int channe
 
+static GstCaps* getGStreamerAudioCaps(float sampleRate, unsigned numberOfChannels)
+{
+    guint64 channel_mask = 0;
```
camelCase please :)

Bastian-Krause replied:
Fixed.

philn reviewed (changes requested):
Thanks, nice clean-up indeed :)

Bastian-Krause committed:
[WebAudio][GStreamer] Use single non-interleaved appsrc instead of multi appsrcs with audiointerleave
https://bugs.webkit.org/show_bug.cgi?id=247736

Reviewed by Philippe Normand.

GStreamer 1.16 brought support for planar (non-interleaved) raw audio.
Let's use that to simplify the "1 appsrc per channel + audiointerleave"
approach to a single appsrc with non-interleaved layout.

We do not need to keep the appsrcs and mapped buffers per channel around
which lets us drop a bunch of code.

This approach cuts the overhead the GStreamer pipeline causes on the
WPEWebProcess in half on an NXP i.MX 6 DualLite.

* Source/WebCore/platform/audio/gstreamer/WebKitWebAudioSourceGStreamer.cpp:
(getGStreamerAudioCaps): Return caps with actual number of channels and
non-interleaved layout.
(webKitWebAudioSrcConstructed): Drop audiointerleave element, create
single appsrc.
(webKitWebAudioSrcAllocateBuffers): Drop loop over multiple sources, no
need to keep mapped buffers per channel around.
(webKitWebAudioSrcRenderAndPushFrames): Drop loop over multiple sources.
(webKitWebAudioSrcRenderIteration): Operate on single buffer.
(webKitWebAudioSrcChangeState): Drop check for audiointerleave element,
adjust buffer size calculation.
(getGStreamerMonoAudioCaps): Deleted. Replaced by getGStreamerAudioCaps.

Canonical link: https://commits.webkit.org/256648@main
